test/unit/org/apache/cassandra/service/PartitionDenylistTest.java

```diff
@@ -0,0 +1,331 @@
+        confirmDenied("iii", "jjj");
+
+        // poke a hole in the middle and reload
+        removeDenylist("" + ks_cql + "", "foofoo", "eee:fff");
+
+        confirmAllowed("eee", "fff");
+        confirmDenied("aaa", "bbb");
+        confirmDenied("iii", "jjj");
+    }
+
+    /**
+     * We go through a few steps here:
+     *  1) Add more keys than we're allowed
+     *  2) Confirm that the overflow keys are *not* denied
+     *  3) Raise the allowable limit
+     *  4) Confirm that the overflow keys are now denied (and no longer really "overflow" for that matter)
+     */
+    @Test
+    public void testShrinkAndGrow() throws InterruptedException
+    {
+        StorageProxy.instance.setMaxDenylistKeysPerTable(5);
+        StorageProxy.instance.setMaxDenylistKeysTotal(5);
+        denyAllKeys();
+
+        // Confirm overflowed keys are allowed; first come first served
+        confirmDenied("aaa", "bbb");
+        confirmAllowed("iii", "jjj");
+
+        // Now we raise the limit back up and do nothing else and confirm it's blocked
+        StorageProxy.instance.setMaxDenylistKeysPerTable(1000);
+        StorageProxy.instance.setMaxDenylistKeysTotal(1000);
+        confirmDenied("aaa", "bbb");
+        confirmDenied("iii", "jjj");
+
+        StorageProxy.instance.setMaxDenylistKeysPerTable(5);
+        StorageProxy.instance.setMaxDenylistKeysTotal(5);
+        confirmAllowed("iii", "jjj");
+
+        // Now, we remove the denylist entries for our first 5, drop the limit back down, and confirm those overflowed keys now block
+        removeDenylist("" + ks_cql + "", "foofoo", "aaa:bbb");
+        removeDenylist("" + ks_cql + "", "foofoo", "bbb:ccc");
+        removeDenylist("" + ks_cql + "", "foofoo", "ccc:ddd");
+        removeDenylist("" + ks_cql + "", "foofoo", "ddd:eee");
+        removeDenylist("" + ks_cql + "", "foofoo", "eee:fff");
+        confirmDenied("iii", "jjj");
+    }
+
+    private void confirmDenied(String keyOne, String keyTwo)
+    {
+        boolean denied = false;
+        String query = String.format("SELECT * FROM " + ks_cql + ".foofoo WHERE bar='%s' and baz='%s'", keyOne, keyTwo);
+        try
+        {
+            process(query, ConsistencyLevel.ONE);
+        }
+        catch (InvalidRequestException ire)
+        {
+            denied = true;
+        }
+        Assert.assertTrue(String.format("Expected query to be denied but was allowed! %s", query), denied);
+    }
+
+    private void confirmAllowed(String keyOne, String keyTwo)
+    {
+        process(String.format("SELECT * FROM " + ks_cql + ".foofoo WHERE bar='%s' and baz='%s'", keyOne, keyTwo), ConsistencyLevel.ONE);
+    }
+
+    private void resetDenylist()
+    {
+        process("TRUNCATE system_distributed.partition_denylist", ConsistencyLevel.ONE);
+        StorageProxy.instance.setMaxDenylistKeysTotal(1000);
+        StorageProxy.instance.setMaxDenylistKeysPerTable(1000);
+        StorageProxy.instance.loadPartitionDenylist();
```

Comment:
I feel this will be a flaky test. `loadPartitionDenylist` reloads data asynchronously, so it is possible that data has not yet been reloaded while the execution goes further. We somehow need to wait here.

smth like `awaitility.await` is required here:
```
await().atMost(5, TimeUnit.SECONDS)
                   .until(() -> <<cache is empty check>>);
```